tests/integration_tests/data_steward/cdr_cleaner/cleaning_rules/cleaning_tests_base.py

```diff
@@ -0,0 +1,224 @@
+        # can only be made in an existing dataset.
+        sandbox_dataset_ids = []
+        # a list of fully qualified table names the cleaning rule is targeting.
+        # fq = {project_id}.{dataset_id}.{table_name}.  required for
+        # test setup and cleanup.
+        fq_table_names = []
+        # a list of fully qualified sandbox tables the rule may create.
+        # required for cleanup
+        fq_sandbox_table_names = []
+        # a list of dictionaries where each dictionary defines table
+        # information for one table upon rule execution, such as the
+        # 'name', 'fq_table_name', 'fq_sanbox_table_name', 'loaded_ids',
+        # 'sandboxed_ids', and 'cleaned_ids'
+        tables_and_counts = []
+        # a list of load statements to execute before each test.  This can be
+        # defined in the extending class's 'setUp(self)' function.
+        sql_load_statements = []
+        # The query class that is being executed.
+        query_class = None
+
+        @classmethod
+        def setUpClass(cls):
+            # get the test project
+            if 'test' not in cls.project_id:
+                raise RuntimeError(
+                    'This should only be run on a test environment')
+
+            # get or create datasets, most cleaning rules assume the datasets exist
+            for dataset_id in cls.test_dataset_ids:
+                bq.get_or_create_dataset(cls.project_id, dataset_id)
+
+            for dataset_id in cls.sandbox_dataset_ids:
+                bq.get_or_create_dataset(cls.project_id, dataset_id)
+
+            # create empty table in test environment
+            cls.client = bq.get_client(cls.project_id)
+
+            # the base class will try to determine this from the list of
+            # tables_and_counts if the list is empty
+            if not cls.fq_table_names:
+                for table_info in cls.tables_and_counts:
+                    cls.fq_table_names.append(
+                        table_info.get('fq_table_name', ''))
+
+            table_info = {
+                # list of fully qualified table names to create.
+                # will be created with project defined schemas
+                'fq_table_names': cls.fq_table_names,
+                'project_id': cls.project_id,
+                'exists_ok': True,
+            }
+            bq.create_tables(**table_info)
+
+        @classmethod
+        def tearDownClass(cls):
+            """
+            Remove the test dataset table(s).
+            """
+            for table in cls.fq_table_names + cls.fq_sandbox_table_names:
+                cls.client.delete_table(table)
+
+        def drop_rows(self, fq_table_name):
+            """
+            Helper function to drop table rows and assert the drop finished.
+
+            Raises an assertion error if the table records are not dropped.
+            Relies on id fields using values greater than -1.
```

Comment:
presumably, primary keys are integers greater than 0.  I used -1 to be safe because this is just for tests.  This is in case someone defines more than one test and the data must be reloaded.  This will drop all rows in the test dataset so that when it is reloaded, it is clean.

Reply:
further expanded this comment.